### ISSUE TITLE
Proposal: Improve match-making routine

### ISSUE THREAD

**Issue by endel** [colyseus/colyseus#176]
I'd like to propose a change in the API, to allow restructuring of the match-making process, and avoid unexpected issues during its process.

The proposed API here is not final. I'd love to hear your thoughts about it!

## Current behaviour

Currently, the match-making process may cause unexpected issues due to async rooms being disposed in the middle of the process, like https://github.com/gamestdio/colyseus/issues/174

On the client-side, the room instance is immediately created even though the room might not be successfully created.

```typescript
const room = client.join("my_room", { /* options */ });
room.onJoin.add(() => console.log("joined successfully!"))
room.onError.add(() => console.log("room couldn't be created?"))
```

- This API can be confusing since both room creation errors, and regular room errors are dispatched into the same handler, without a differentiation between them.
- It's not possible to determine whether a join error was caused by wrong credentials (`onAuth`) or a bug in the server.

## Desired behaviour

The server can make extra-checks during the match-making process to avoid this error (https://github.com/gamestdio/colyseus/issues/174), and provide better error handling in the client-side.

```typescript
client.matchmake("my_room", {/* options */}, (err, room) => {
  if (err) { 
    console.log("match-making error: couldn't join into 'my_room'");
    console.log("reason:", err)
    return;
  }

  room.onJoin.add(() => console.log("joined successfully!"));
  room.onError.add(() => console.log("only room errors here"));
});
```

## Version `1.0.0`

As this is a big change in the API, I'm planning to have this on version `1.0.0`.

**Comment by seiyria** [comment]
Anything to fix #174 sounds great to me! If the `matchmake` function can resolve this, I'm fine with it.

**Comment by AnubisCode** [comment]
good idea!

**Comment by seiyria** [comment]
I should add: will this _also_ fire on reconnect? I think it would need to otherwise the same issues as previous will crop up on a reconnect.

**Comment by vapurrmaid** [comment]
looks like a good plan

**Comment by oyed** [comment]
I'm all for this as well as potentially separating Matchmaking from Colyseus entirely.

**Comment by seiyria** [comment]
@oyed I'd like to make it optional - I certainly don't want to deal with hosting an external framework/tool/service just for matchmaking (I don't even use matchmaking since my game is an rpg, I just use the routing - though I am admittedly in the minority here :D). Optionally would be good!

**Comment by oyed** [comment]
@seiyria I'm not entirely sure you're in the minority - Colyseus is a netcode lib, not a matchmaking lib - attempting to do "too much" could easily take away what Colyseus is, a netcode lib. It would be nice to optionally have a built-in, simplistic, non-scalable matchmaking service like we currently do, however for the future, supporting something like mentioned in #179 would really benefit the library and allow @endel to focus on more important things.

**Comment by gioragutt** [comment]
@oyed why not just extract all the matchmaking logic to `@gamestdio` and have it like interact in a pluggable way `colyseus`?

**Comment by oyed** [comment]
@gioragutt Sure, that'd allow for pluggable matchmaking with a simple colyseus-defined interface class.

**Comment by endel** [comment]
There are many issues (#205, #200, #201, #196, #183) open regarding the current "presence" mechanism, using either `MemsharedPresence` or `RedisPresence`. Isolating the matchmaking completely would make a lot of sense. I'm going to expose my thoughts on this.

### Where we are:

The major issue is that a client - once connected into a Room - may be sending messages to one process, but the actual Room he's connected to lives on another process. These messages need to be passed back-and-forth between Node processes, using the `presence` mechanism.

Apparently **it is** possible to use sticky-sessions if using `pm2` + `nginx` (https://github.com/Unitech/pm2/issues/389), but this wouldn't solve the need for communication between processes completely, only for the first client creating the room. 

### Where should we aim for?

- A separation of **Matchmaking** and **Game Server** would be necessary
    - **Matchmaking** processes never spawn Rooms
    - **Game Server** spawn Rooms and report back to Matchmaking
- It should be possible to continue using Matchmaking + Game Servers in the same Node.js app without extra configuration (for prototyping)
- It should be possible to have multiple Matchmaking servers/instances
- Matchmaking instances tells the client which Game Server he's going to connect
- Each Game Server will need to have a public way to access it (e.g. `srv1.mygame.io/process1`, `srv2.mygame.io/process4`, etc., this means extra configuration)
- When connected to the Game Server, the client communicate only to the Node instance which his Room lives.
- Ideally easy to configure, but will definitely increase steps & configuration for scaling

This may all sound obvious but it's not easy to implement. I'd love to hear your thoughts if you have any advice or suggestion to make!

Cheers! Enjoy the holidays everyone!

**Comment by LightningK0ala** [comment]
I'm in favor. Can we expose the matchmaker with graphql?

**Comment by seiyria** [comment]
Graphql is a data query layer, not a logic layer.

**Comment by ClosetMonkey** [comment]
ZeroMQ handles such cases by queuing events when they cannot be sent and then to draing the queue when expected connection comes back up (or if a different process/endpoint steps in to collect the events).

**Comment by endel** [comment]
Hi everyone! I'm working on version `0.11.0` this week, and I really wanna improve matchmaking/scalability.

**TLDR: the `requestJoin()` method is going to be removed soon!**

The main reason why the current matchmaking doesn't scale well is because of the `requestJoin()` method.

Let's take my game ([mazmorra.io](https://mazmorra.io)) as an example. I have a `"dungeon"` room, which can be spawned with a `progress` option, to determine how deep the player are in the dungeons.

**index.ts**

```typescript
gameServer.register("dungeon", DungeonRoom); // regular dungeon room
```

**DungeonRoom.ts**

```typescript
// ...
onInit(options) {
  this.progress = options.progress;
}

requestJoin (options, isNew) {
  return (options.progress === this.progress);
}
// ...
```

The problem is, let's say there are 20 players online - each one on different dungeon progress. When a player requests to switch between his current `progress` to a new one, the matchmaker will perform a `requestJoin()` on **each of the other progresses**. This is simply not acceptable.

So my proposal is:

## 1. Removing the `requestJoin()` method

The `requestJoin()` method exists [since the very first version of Colyseus](https://github.com/colyseus/colyseus/blob/6c3814eefb1d11ad062e7a8e09673c7d31c5fd60/lib/room.js#L17), and it turns out it was a bad design decision. We cannot rely on the instance scope of the room to determine if players can join it or not. We need a separate structure to match against without the cost of asking the machine/process where the room instance is located to respond.

**Example:**

This example defines a `"dungeon"` room type, with a default `progress = 1` option, and determines it is filtered by the `"progress"` option for matchmaking.

```typescript
gameServer.
    define("dungeon", DungeonRoom, { progress: 1 }).
    filterBy(['progress']);
```

All the rooms will be registered in a database (mongodb probably), and removed from it upon disposal - with the provided `filterBy` fields for being searchable.

## 2. More control to the client-side

After this change, these methods will be available in the client-side:
- `joinOrCreate(roomName, options)` - joins or creates a room by name 
- `create(roomName, options)` - only creates new rooms
- `join(roomName, options)` - only joins existing rooms by name
- `joinById(roomName, options)` - only joins existing rooms by id

## Feedback? 👀 

Your feedback is always welcome, specially regarding the public API for `gameServer.define(...).filterBy()` which I'm not sure the naming convention is good or not.

Thanks for reading until here. Cheers!

**Comment by damian-pastorini** [comment]
@endel I really like these improvements, but once you change the client and split the current single join in those multiple methods (joinOrCreate, join, create, which will be much clear for a lot of people who ask about how to handle the current single join), will be possible to "control" or maybe will be a good idea to include a way to validate if the client can really create a room or just join a room?

**Comment by Zielak** [comment]
1. In that example, if I just hack client to send `{ progress: 100 }`, the server would still require to perform a check, if player has access to that level. Where would you put such logic with out `requestJoin()`?

2. Taking your game as an example again, if a player walks in to the "next level" door (sends message { command: "nextLevel" }), shouldn't the server now be responsible to perform all the checks (*is player near door?*) and push the player into the next level/room? I imagine that behavior would be ideal.

I'm starting to think that `requestJoin()` still has its uses, but given that its not performant right now, it could be called on the last step:

1. client: `join({prorgess: 2})`
2. server: `filterBy('progress') `
3. server: finds all rooms where `progress === 2`
4. server: calls `requestJoin` for all the filtered out rooms.
    - maybe we have multiple "progress: 2" rooms, so one level wouldn't get overcrowded.

**Comment by Roxgame95** [comment]
Nice waiting for new release join and create now more clear

**Comment by endel** [comment]
Hey @Zielak, thanks for your input, if the user tries to "hack" by sending options he shouldn't be sending, you can validate them during `onAuth()` or `onJoin()`, and throw an Error there - which will prevent him from joining.

The `requestJoin(options, isNew)` method wasn't really useful for checking the scenario you described, because no user data was being provided to it besides the join options itself. 

The alpha version of `0.11` is already available on NPM under the tag `colyseus@alpha` / `colyseus.js@alpha`. I still need to check how to handle private password-protected rooms with the new matchmaking structure, so far everything is looking simpler and easier to maintain, a lot of complexity is going to be removed from the client-side implementations as well!

**Comment by endel** [comment]
@Zielak about the priority for joining rooms, so far I've came up with this API:

```typescript
gameServer.define("dungeon", DungeonRoom).
  filterBy(['progress']).
  sortBy({ clients: -1 });
```

Using this definition, when receiving a join call for `"dungeon"`, a query will run on the database sorting by a descending number of connected clients. 

You can provide any field stored in the database for `sortBy()`. The `filterBy()` options are going to be stored in the database as well (if any). Values such as `clients` and `maxClients` are stored by default.

I'm not 100% confident this is going to be a great solution, maybe this is going to change in the future

**Comment by endel** [comment]
Version `0.11` is ready! 🎉 
See the migration guide for 0.11 here: https://docs.colyseus.io/migrating/0.11/
